pkg/pgmodel/pgx.go

```diff
@@ -1021,6 +1025,84 @@ func (q *pgxQuerier) LabelValues(labelName string) ([]string, error) {
 	return labelValues, nil
 }
 
+const GetLabelsSQL = "SELECT (key_value_array($1::int[])).*"
+
+type labelQuerier interface {
+	getLabelsForIds(ids []int64) (lls labels.Labels, err error)
+}
+
+func (q *pgxQuerier) getPrompbLabelsForIds(ids []int64) (lls []prompb.Label, err error) {
+	ll, err := q.getLabelsForIds(ids)
+	if err != nil {
+		return
+	}
+	lls = make([]prompb.Label, len(ll))
+	for i := range ll {
+		lls[i] = prompb.Label{Name: ll[i].Name, Value: ll[i].Value}
+	}
+	return
+}
+
+func (q *pgxQuerier) getLabelsForIds(ids []int64) (lls labels.Labels, err error) {
+	keys := make([]interface{}, len(ids))
+	values := make([]interface{}, len(ids))
+	for i := range ids {
+		keys[i] = ids[i]
+	}
+	numHits := q.labels.GetValues(keys, values)
+
+	if numHits < len(ids) {
+		err = q.fetchMissingLabels(keys[numHits:], ids[numHits:], values[numHits:])
+		if err != nil {
+			return
+		}
+	}
+
+	lls = make([]labels.Label, 0, len(values))
+	for i := range values {
+		lls = append(lls, values[i].(labels.Label))
+	}
+
+	return
+}
+
+func (q *pgxQuerier) fetchMissingLabels(misses []interface{}, missedIds []int64, newLabels []interface{}) error {
+	for i := range misses {
+		missedIds[i] = misses[i].(int64)
+	}
+	rows, err := q.conn.Query(context.Background(), GetLabelsSQL, missedIds)
+	if err != nil {
+		return err
+	}
+
+	for rows.Next() {
+		var keys []string
+		var vals []string
+		err = rows.Scan(&keys, &vals)
+		if err != nil {
+			return err
+		}
+		if len(keys) != len(vals) {
+			return fmt.Errorf("query returned a mismatch in timestamps and values: %d, %d", len(keys), len(vals))
+		}
+		if len(keys) != len(misses) {
+			return fmt.Errorf("query returned wrong number of labels: %d, %d", len(misses), len(keys))
+		}
+
+		lls := Labels{names: keys, values: vals}
+		sort.Sort(&lls)
+
+		for i := range misses {
+			newLabels[i] = labels.Label{Name: lls.names[i], Value: lls.values[i]}
+		}
```

Comment:
```suggestion
		for i := range keys {
			newLabels[i] = labels.Label{Name: keys[i], Value: vals[i]}
		}
```

Reply:
I have mixed feelings: on the one hand the resulting code is smaller, on the other hand once the Labels take ownership of the arrays it seems better to always access it through that